packages/dataviews/src/components/dataform/stories/index.story.tsx

```diff
@@ -1261,3 +1278,211 @@ export const Validation = {
+			readOnly: true,
+		},
+		{
+			id: 'revenue.pricePerUnit',
+			label: 'Price Per Unit',
+			type: 'integer',
+			setValue: ( { item, value } ) => ( {
+				revenue: {
+					total: value * item.revenue.units,
+					pricePerUnit: value,
+				},
+			} ),
+		},
+		{
+			id: 'revenue.units',
+			label: 'Units',
+			type: 'integer',
+			setValue: ( { item, value } ) => ( {
+				revenue: {
+					total: item.revenue.pricePerUnit * value,
+					units: value,
+				},
+			} ),
+		},
+	];
+
+	const handleChange = useCallback( ( edits: any ) => {
+		// Edits will respect the shape of the data
+		// because fields provide the proper information
+		// (via field.id or via field.setValue).
+		setData( ( prev ) => deepMerge( prev, edits ) );
+	}, [] );
+
+	return (
+		<>
+			<h1>Data adapter</h1>
+			<p>
+				This story is best looked at with the code on the side. It aims
+				to highlight how DataForm can wrangle data in scenarios such as
+				nested data, bridge data to/from UI controls, and derived data.
+			</p>
+			<p>
+				<b>Current data snapshot:</b>
+			</p>
+			<pre>{ JSON.stringify( data, null, 2 ) }</pre>
+			<h2>Nested data</h2>
+			<p>
+				The first example demonstrates how to signal nested data via{ ' ' }
+				<code>field.id</code>.
+			</p>
+			<p>
+				By using <code>{ `{ id: 'user.profile.name' }` }</code> as field
+				id, when users edit the name, the edits will come in this shape:
+				<code>{ `{ user: { profile: { name: 'John Doe' } } }` }</code>
+			</p>
+			<DataForm< DataAdapterItem >
+				data={ data }
+				fields={ nestedFields }
+				form={ {
+					layout: {
+						type: 'panel',
+						labelPosition: 'top',
+						openAs: 'modal',
+					},
+					fields: [
+						{
+							id: 'userProfile',
+							label: 'User Profile',
+							children: [
+								'user.profile.name',
+								'user.profile.email',
+							],
+						},
+					],
+				} }
+				onChange={ handleChange }
+			/>
+			<h2>Adapt data and UI control</h2>
+			<p>
+				Sometimes, we need to adapt the data type to and from the UI
+				control response. This example demonstrates how to adapt a
+				boolean to a text string (Enabled/Disabled).
+			</p>
+			<DataForm< DataAdapterItem >
+				data={ data }
+				fields={ nestedFields }
+				form={ {
+					layout: {
+						type: 'panel',
+						labelPosition: 'top',
+						openAs: 'modal',
+					},
+					fields: [ 'user.preferences.notifications' ],
+				} }
+				onChange={ handleChange }
+			/>
+			<h2>Derived data</h2>
+			<p>
+				Last, but not least, this example showcases how to work with
+				derived data by providing a custom setValue function. Note how,
+				changing UNITS or PRICE PER UNIT, updates the TOTAL value as
+				well.
+			</p>
+			<DataForm< DataAdapterItem >
+				data={ data }
+				fields={ nestedFields }
+				form={ {
+					layout: {
+						type: 'panel',
+						labelPosition: 'top',
+						openAs: 'modal',
+					},
+					fields: [
+						{
+							id: 'revenue',
+							label: 'Revenue',
+							children: [
```

Comment:
This example would be improved when [this PR](https://github.com/WordPress/gutenberg/pull/71614/) lands, and we can set a summary (revenue.total).